### PR TITLE
Convert BsButton to TypeScript

### PR DESCRIPTION
**Pull request by simonihmig**
Converts our "leaf-most" component to TS, including type checking of its template and emitting declarations.

When reviewing, check [this commit](https://github.com/ember-bootstrap/ember-bootstrap/pull/2093/commits/e641f84b77f8ca932ad0c0e386a9f26c092b8912) specifically, as it builds on top of #2092 to tweak the tsconfig such that emitting declarations with Glint actually work...

### CODE REVIEW

File: .vscode/settings.json
Comment (simonihmig) [inline .vscode/settings.json:2]: The Glint VSCode plugin expect to find the glint libraries in the root's node_modules, where they aren't due to pnpm not hoisting dependencies. So this is making the plugin work ootb when setting up the repo locally. Normally, I wouldn't but code editor configs into the repo, but in this case maybe it makes sense?

**Comment by simonihmig** [comment]
When reviewing, check [this commit](https://github.com/ember-bootstrap/ember-bootstrap/pull/2093/commits/e641f84b77f8ca932ad0c0e386a9f26c092b8912) specifically, as it builds on top of #2092 to tweak the tsconfig such that emitting declarations with Glint actually work...

**Review by jelhan — CHANGES_REQUESTED** [review]
I feel we should convert the corresponding tests to TypeScript as well. Without that we don't have test coverage for the types, which makes reviewing and maintaining the type signatures more difficult.

**Comment by jelhan** [comment]
> I feel we should convert the corresponding tests to TypeScript as well. Without that we don't have test coverage for the types, which makes reviewing and maintaining the type signatures more difficult.

I started converting `<BsAccordion>` to TypeScript. Including the tests. And noticed late that the test app is not setup for TypeScript yet. I feel we should do that as another preparation step before.

**Comment by simonihmig** [comment]
I agree in general, but I wonder if we need to have everything in place in a single PR. I think we will only be able to call the types to be safe and stable when we have converted _everything_, all components and all tests. Like a future PR that converts ButtonGroup for example could reveal that some (internal) button types were missing or wrong. But making these changes land sooner than later unblocks those other migrations, as we have to convert basically from the leaf components upwards (or downwards, however you think of the orientation of that tree :wink:). So not sure if we need to hold back iterative improvements like this one?

**Comment by jelhan** [comment]
> I agree in general, but I wonder if we need to have everything in place in a single PR.

I don't think we need to have everything in one single PR. Actually I think we should split it into small chunks. I just have bad experience from splitting feature implementation and tests for it in separate PRs.

I see that there could be edge cases, in which converting the corresponding tests with the component to TypeScript may not be easy. E.g. if the tests render another component, which has not been converted yet for testing compatibility. I agree that we need to be pragmatic in those cases. But that doesn't seem to be the case for `<BsButton>`.

@SanderKnauff already created a PR converting the test app to TypeScript. This should unblock converting corresponding types together with the components. I hope I find the time later today pulling it into #2095. Getting some real-world experience on overhead due to converting tests as well, will likely help us with decision making.

**Comment by simonihmig** [comment]
Just merged #2096, so will update this PR to have the tests covered as well!

**Comment by jelhan** [comment]
> Just merged #2096, so will update this PR to have the tests covered as well!

Have a look at #2095. I converted some test infrastructure, which you may need as well, to TypeScript already. Especially the `test-app/tests/helpers/bootstrap`. Not sure if the types are correct. I only done minimal work to get it passing.

**Comment by simonihmig** [comment]
> I converted some test infrastructure, which you may need as well, to TypeScript already.

Oh no, just did that as well! 🙈 Should looked at your work first 😅

**Review by jelhan — APPROVED** [review]
Looks good to me.

**Comment by simonihmig** [comment]
CI is finally happy.! :tada:

Feel free to have a final look, but I hope it's good now! Keeping this open for a bit, in case @SanderKnauff you want to also review before merging. But besides the actual button-related migration, this also contains a bit of infrastructure (test-helpers, CI/scripts setup) that is needed for follow-up migrations, so feel free to hit the merge button when anyone needs this to unblock their work! :)

File: ember-bootstrap/addon/components/bs-button.hbs
Comment (SanderKnauff) [inline ember-bootstrap/addon/components/bs-button.hbs:4]: While this does work, I think the newlines will be included while rendering.
Is this something we want to look out for?

File: ember-bootstrap/addon/utils/decorators/arg.ts
Comment (SanderKnauff) [inline ember-bootstrap/addon/utils/decorators/arg.ts:10]: With generics, is it possible to type args like `<T extends object>(target: T, key: keyof T)`?

**Review by SanderKnauff — CHANGES_REQUESTED** [review]
Awesome start! Just a few small comments.
I do hope that after the initial boilerplate work is done, the changesets will become a bit more manageable.

File: ember-bootstrap/addon/utils/decorators/arg.ts
Comment (simonihmig) [reply ember-bootstrap/addon/utils/decorators/arg.ts:10]: IIRC I tried that, but at some point I gave up. `key` has to be a keyof this.args actually, see L23. But given this is used as a decorator, which has its own expected signature (with TS actually disagreeing on our implementation), this was not working easily, or I was missing something...

File: ember-bootstrap/addon/components/bs-button.hbs
Comment (simonihmig) [reply ember-bootstrap/addon/components/bs-button.hbs]: I just fixed that!

I think our prettier setup is not right, wanted to come up with a PR to fix this. With pnpm and no hoisting, we don't have prettier in the root node_modules, but VS Code expects it there to work with the prettier plugin. I believe as this was not working ootb, it would reformat the code, but not based on prettier. Had to run `lint:fix` sometimes, which is unexpected, as usually prettier-in-the-editor should be in-sync with prettier-in-lint-scripts.

**Comment by simonihmig** [comment]
> I do hope that after the initial boilerplate work is done, the changesets will become a bit more manageable.

Sure! 😅

File: ember-bootstrap/addon/components/bs-button.hbs
Comment (SanderKnauff) [reply ember-bootstrap/addon/components/bs-button.hbs]: hmm, should the prettier configuration be in the root project?

That aside, I did like the indentation as it was before, it's just a shame that it generates all those newlines. Maybe we can add a helper in the future that strips all the unneccesary whitespaces from such class lists.

File: ember-bootstrap/addon/utils/decorators/arg.ts
Comment (SanderKnauff) [reply ember-bootstrap/addon/utils/decorators/arg.ts:10]: I'm not that familiar with creating decorators, so I believe you if you say that it's not working. :sweat_smile:

**Review by SanderKnauff — APPROVED** [review]

File: ember-bootstrap/addon/components/bs-button.hbs
Comment (jelhan) [reply ember-bootstrap/addon/components/bs-button.hbs]: Having a build plugin which strips whitespaces would be great. But I'm not sure if something like that already exists in Ember ecosystem.

Comment (simonihmig) [reply ember-bootstrap/addon/components/bs-button.hbs]: There is this: https://github.com/mainmatter/ember-hbs-minifier 

But not sure if it's able to only run on our addon. Also when migrating to v2, we would want to have something that runs ahead of time as part of our rollup build...

Comment (jelhan) [reply ember-bootstrap/addon/components/bs-button.hbs]: Yeah. I have more thinking in a v2 addon world. I feel it's too risky and unnecessary churn introducing such tooling before v2 conversion.

**Pull request merged**
